distributed/scheduler.py

```diff
@@ -3596,6 +3596,84 @@ def transition_no_worker_waiting(self, key):
                 pdb.set_trace()
             raise
 
+    def decide_worker_initial(self, ts: TaskState) -> WorkerState:
+        r"""
+        Decide the worker to assign a task to.
+
+        Parameters
+        ----------
+        ts : TaskState
+            This is a ready task that has no dependencies.
+
+        Returns
+        -------
+        WorkerState
+
+        Notes
+        -----
+        This prioritizes scheduling a task on a worker executing tasks that we're
+        a sibling dependency with.
+
+        Consider the following task graph (read top to bottom)
+
+        ::
+
+            a-1  a-2  a-3  a-4
+             \   /     \   /
+              b-1       b-2
+
+        If we have
+
+        * Two workers: `A` and `B`
+        * Task `a-1` is running on worker `A`
+        * We're currently scheduling task `a-2`
+
+        we'll choose to schedule `a-2` it on worker `A` as well because that will
+        minimize communication when `b-1` is eventually scheduled.
+
+        When we dont have any sibling tasks running, we assign `ts` to an idle worker,
+        or a worker with occupancy / a relatively low number of tasks.
+
+        See Also
+        --------
+        decide_worker
+        worker_objective
+        transition_waiting_processing
+        """
+        worker = None
+        # the config is just for ease of testing / benchmarking. Will remove
+        if dask.config.get("distributed.scheduler.lump_tasks", default=True):
+            for dts in ts.dependents:
+                # Figure out where my siblings are running. Note that we
+                # stop as soon as we find *a* sibling running *somewhere*.
+                # If time weren't an issue, we might find the worker with the
+                # most siblings. But that's expensive.
+                #
+                for sts in dts.dependencies:
```

Comment:
There are several situations where a single task has very many dependents.  In these cases I think that we'll hit N^2 scaling and bring things down.

Reply:
What about cases where we don't have siblings, but cousins n'th removed

```
a1
|
a2    b1
 \   /
   c
```

Reply:
Yes, my initial got a full count of where each of our co-dependencies was running. That blew up very quickly. The early `break` once we find a co-dependency was a first attempt to avoid that.

Reply:
This approach won't help in that case (I think `a-1` and `b-1` are niblings 😄).